zarr/core.py

```diff
@@ -2353,13 +2353,18 @@ def hexdigest(self, hashname="sha1"):
 
         return checksum
 
-    def __getstate__(self):
-        return (self._store, self._path, self._read_only, self._chunk_store,
-                self._synchronizer, self._cache_metadata, self._attrs.cache,
-                self._partial_decompress, self._write_empty_chunks, self._version)
-
-    def __setstate__(self, state):
-        self.__init__(*state)
+    def __reduce__(self):
```

Comment:
Thoughts on implementing `__getnewargs_ex__` instead?

In particular [this Python doc]( https://docs.python.org/3/library/pickle.html#pickling-class-instances ) text seems relevant:

> Although powerful, implementing `__reduce__()` directly in your classes is error prone. For this reason, class designers should use the high-level interface (i.e., `__getnewargs_ex__()`, `__getstate__()` and `__setstate__()`) whenever possible.

Same question with other changes below

Reply:
how does implementing `__getnewargs_ex__` avoid the need to call `__init__` directly in `__setstate__`?

Reply:
Because this only return the arguments that are passed to `__new__` and not a function to call (nor does it call one) itself. Also because `__setstate__` would be dropped then.

Reply:
Note: We may be able to use `__getnewargs__` instead of `__getnewargs_ex__` if we only use positional arguments, which is preferred

From [`__getnewargs_ex__` docs]( https://docs.python.org/3/library/pickle.html#object.__getnewargs_ex__ ):

> You should implement this method if the [`__new__()`]( https://docs.python.org/3/reference/datamodel.html#object.__new__ ) method of your class requires keyword-only arguments. Otherwise, it is recommended for compatibility to implement [`__getnewargs__()`]( https://docs.python.org/3/library/pickle.html#object.__getnewargs__ ).

Reply:
`__new__` is run before `__init__`. The arguments would go to `__init__` as well. We just wouldn't have to do that ourselves. Details in [these docs]( https://docs.python.org/3/reference/datamodel.html#object.__new__ ).

Sure this would work

```python
import pickle


class Value:
    def __init__(self, value):
        self.value = value

    def __getnewargs__(self):
        return (self.value,)


v = Value(5)
v2 = pickle.loads(pickle.dumps(v))
assert 5 == v.value == v2.value
```

Reply:
OK, that's very helpful! I'll look into using this instead of `__reduce__`